rfcs/20181016-replicator.md

```diff
@@ -0,0 +1,962 @@
+  """An input iterator, intended to be passed to `Replicator.run`."""
+
+  def get_next(self) -> List[T]:
+    """Returns the next inputs for each replica."""
+
+  def reinitialize(self) -> List[tf.Operation]:
+    """Re-initializes the inputs."""
+```
+
+### Model Parallelism
+
+Model parallelism can allow the execution of models when there is insufficient
+memory on a single device (and further data parallelism is not possible).
+
+Replicator supports model partitioning, a form of model parallelism in which the
+ops within the model are placed onto multiple devices. This is done via the
+`logical_device` context.
+
+Inputs are always located on logical device 0. Ops within the step function
+default to logical device 0 (except on TPU, which defaults to automatic
+placement).
+
+```python
+with replicator.scope():
+  with replicator.logical_device_for_variables(1):
+    v = tf.get_variable(...)  # On logical device 1.
+    x = v * 3  # Ignores logical device context.
+
+def step(ctx, inputs):  # `inputs` on logical device 0.
+  a = model_part1(inputs)  # Implicitly on logical device 0.
+  with ctx.logical_device(0):
+    b = model_part2(a)  # Explicitly on logical device 0.
+  with ctx.logical_device(1):
+    return model_part3(b)  # Explicitly on logical device 1.
+```
+
+### Eager mode support
+
+Eager mode support mostly drops out of building atop of `DistributionStrategy`.
+The entire API is designed to be eager compatible. We've called out the
+difference in behavior in eager and graph mode in the specific API
+documentations. In general, in graph mode, many of the APIs return ops which
+should be later executed with `session.run`, and in eager mode, they simply run
+the ops.
+
+## Usage Examples
+
+### Building a Replicator object
+
+The first step is to build a `Replicator` instance for the chosen platform:
+
+```python
+dist_strat = ...  # Create `DistributionStrategy` for a specific regime.
+replicator = tf.replicator.Replicator(dist_strat)
+```
+
+### Simple Classification
+
+Below is a simple usage example for an image classification use case.
+
+#### Training
+
+```python
+with replicator.scope():
+  model = resnet.ResNetV1(resnet.BLOCKS_50)
+  optimizer = tf.train.MomentumOptimizer(learning_rate, 0.9)
+
+def input_fn(ctx):
+  assert effective_batch_size % ctx.num_replicas_in_sync == 0
+  return imagenet.ImageNet(effective_batch_size // ctx.num_replicas_in_sync)
+
+def step_fn(ctx, inputs):
+  del ctx  # Unused.
+  image, label = inputs
+  logits = model(images)
+  cross_entropy = tf.nn.sparse_softmax_cross_entropy_with_logits(
+      logits=logits, labels=label)
+  loss = tf.reduce_mean(cross_entropy)
+  train_op = optimizer.minimize(loss)
+  with tf.control_dependencies([train_op]):
+    return tf.identity(loss)
+
+inputs = replicator.prepare_input(input_fn)
+per_replica_losses = replicator.run(step_fn, inputs)
+mean_loss = tf.reduce_mean(tf.stack(per_replica_losses))
+
+with tf.Session() as session:
```

Comment:
How does this work in 2.0?

Reply:
The Reinforcement Learning example is written in eager style. Basically, calling `run` will run the computation instead of returning the op, so we just call `run` inside a loop.